src/main/java/org/breedinginsight/daos/TraitDAO.java

```diff
@@ -181,6 +202,130 @@ public TraitDAO(Configuration config, DSLContext dsl) {
+                    .categories(trait.getScale().getCategories())
+                    .max(trait.getScale().getValidValueMax())
+                    .min(trait.getScale().getValidValueMin())
+                    .build();
+            BrApiScale brApiScale = BrApiScale.builder()
+                    .scaleName(trait.getScale().getScaleName())
+                    .externalReferences(List.of(scaleReference))
+                    .dataType(brApiTraitDataType)
+                    .decimalPlaces(trait.getScale().getDecimalPlaces())
+                    .validValues(brApiScaleValidValues)
+                    .build();
+
+            // Construct trait
+            BrApiExternalReference traitReference = BrApiExternalReference.builder()
+                    .referenceID(trait.getId().toString())
+                    .referenceSource(referenceSource)
+                    .build();
+            BrApiTrait brApiTrait = BrApiTrait.builder()
+                    .traitName(trait.getTraitName())
+                    .traitDescription(trait.getDescription())
+                    .synonyms(trait.getSynonyms())
+                    .status("active")
+                    .entity(trait.getProgramObservationLevel().getName())
+                    .mainAbbreviation(trait.getMainAbbreviation())
+                    .alternativeAbbreviations(trait.getAbbreviations() != null ? List.of(trait.getAbbreviations()) : null)
+                    .traitClass(trait.getTraitClass())
+                    .externalReferences(List.of(traitReference))
+                    .attribute(trait.getAttribute())
+                    .build();
+
+            BrApiExternalReference variableReference = BrApiExternalReference.builder()
+                    .referenceID(trait.getId().toString())
+                    .referenceSource(referenceSource)
+                    .build();
+            BrApiVariable brApiVariable = BrApiVariable.builder()
+                    .method(brApiMethod)
+                    .scale(brApiScale)
+                    .trait(brApiTrait)
+                    .externalReferences(List.of(variableReference))
+                    .observationVariableName(trait.getTraitName())
+                    .status("active")
+                    .language("english")
+                    .scientist(actingUser.getName())
+                    .defaultValue(trait.getDefaultValue())
+                    .synonyms(trait.getSynonyms())
+                    .institution(program.getName())
+                    .commonCropName(program.getSpecies().getCommonName())
+                    .build();
+
+                    // Unused
+                    //.contextOfUse()
+                    //.documentationURL()
+                    //.growthStage()
+
+            brApiVariables.add(brApiVariable);
+        }
+
+
+        // POST variables to each brapi service
+        // TODO: If there is a failure after the first brapi service, roll back all before the failure.
```

Comment:
Do we have a card for api roll back and more specifically applying it here when we have it?

Reply:
https://trello.com/c/sZmjsdYW/37-acid-transactions-for-brapi-calls